crates/virtio-queue/src/state_sync.rs

```diff
@@ -78,26 +85,207 @@ impl QueueStateT for QueueStateSync {
         head_index: u16,
         len: u32,
     ) -> Result<(), Error> {
-        self.state.lock().unwrap().add_used(mem, head_index, len)
+        self.lock_state().add_used(mem, head_index, len)
     }
 
     fn enable_notification<M: GuestMemory>(&mut self, mem: &M) -> Result<bool, Error> {
-        self.state.lock().unwrap().enable_notification(mem)
+        self.lock_state().enable_notification(mem)
     }
 
     fn disable_notification<M: GuestMemory>(&mut self, mem: &M) -> Result<(), Error> {
-        self.state.lock().unwrap().disable_notification(mem)
+        self.lock_state().disable_notification(mem)
     }
 
     fn needs_notification<M: GuestMemory>(&mut self, mem: &M) -> Result<bool, Error> {
-        self.state.lock().unwrap().needs_notification(mem)
+        self.lock_state().needs_notification(mem)
     }
 
     fn next_avail(&self) -> u16 {
-        self.state.lock().unwrap().next_avail()
+        self.lock_state().next_avail()
     }
 
     fn set_next_avail(&mut self, next_avail: u16) {
-        self.state.lock().unwrap().set_next_avail(next_avail);
+        self.lock_state().set_next_avail(next_avail);
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Hmm, looks like most of the tests here would apply to other `QueueStateT` implementations just as well. It would actually be great to have a battery of tests for the `QueueStateT` interface that can then be reused to test all implementations (so far there's `QueueState` and `QueueStateSync`), and then separate tests for stuff that's specific to `QueueStateSync` in particular (altho there's not that much truly specific stuff IMO; even the test where multiple threads are spawned can be seen as generic due to `lock` being part of the high level interface).

Not sure exactly just how much effort that would require, so maybe doing it as part of this PR is not the best approach, but then I think we should at least have an issue and tackle it b4 publishing.

Reply:
Will open one issue for it once the PR get merged:)